### PR TITLE
[TOPI] bitserial_conv2d move to autotvm template and updates

### PR DESCRIPTION
**Pull request by cowanmeg**
Updates to arm_cpu and x86 bitserial_conv2d operators
- Moved to autotvm template
- Support kernels that are pre-bitpacked for arm_cpu

### CODE REVIEW

**Comment by cowanmeg** [comment]
@ZihengJiang

**Comment by tqchen** [comment]
Please request reviewers

**Comment by cowanmeg** [comment]
@eqy @ZihengJiang, please review

File: topi/python/topi/arm_cpu/bitserial_conv2d.py [hunk 1]
```diff
@@ -1,331 +1,315 @@
...
+                              filter=lambda x: x.size[-1] >= 2)
+    # cfg.define_annotate('ann_reduce', [ib, kb, kh, kw], policy='try_unroll')
+    ci_o, ci_i = cfg.define_split("tile_ci", ci, num_outputs=2,
+                                  filter=lambda x: x.size[-1] == 8 or x.size[-1] == 16)
+    re_axes = cfg.define_reorder("reorder_0",
+                                 [n, oh, ow, co, vh, vw, kh, kw, ci_o, kb, ib, vc, ci_i],
+                                 policy='candidate', candidate=[
+                                     [n, oh, ow, co, vh, vw, kh, kw, ci_o, kb, ib, vc, ci_i],
+                                     [n, oh, ow, co, vh, vw, kw, kh, ci_o, kb, ib, vc, ci_i],])
+    cfg.add_flop(2 * N * OH * OW * CO * CI * 8 * KH * KW)
```
Comment (eqy) [inline topi/python/topi/arm_cpu/bitserial_conv2d.py]: Does the automatic flop calculation in #2776 work for this?

File: topi/python/topi/nn/bitserial_conv2d.py [hunk 2]
```diff
@@ -203,40 +298,70 @@ def _conv(n, co, h, w, vh, vw, vc):
...
+                              filter=lambda x: max(x.size[1:]) <= 16)
+    oh, vh = cfg.define_split('tile_oh', oh, policy='all', num_outputs=2,
+                              filter=lambda x: max(x.size[1:]) <= 16)
+    ow, vw = cfg.define_split('tile_ow', ow, policy='all', num_outputs=2,
+                              filter=lambda x: max(x.size[1:]) <= 16)
+    cfg.define_annotate('ann_reduce', [ib, kb, kh, kw], policy='try_unroll')
+    re_axes = cfg.define_reorder("reorder_0",
+                                 [n, oh, ow, co, vh, vw, kh, kw, kb, ib, vc, ci],
+                                 policy='interval_all', interval=(3, 7))
+    cfg.add_flop(2 * N * OH * OW * CO * CI * 8 * KH * KW)
```
Comment (eqy) [inline topi/python/topi/nn/bitserial_conv2d.py]: ""

File: topi/python/topi/nn/bitserial_conv2d.py [hunk 3]
```diff
@@ -78,102 +40,235 @@ def bitserial_conv2d(data, kernel, stride, padding, activation_bits, weight_bits
...
+    oh, vh = cfg.define_split('tile_oh', oh, policy='all', num_outputs=2,
+                              filter=lambda x: max(x.size[1:]) <= 16)
+    ow, vw = cfg.define_split('tile_ow', ow, policy='all', num_outputs=2,
+                              filter=lambda x: max(x.size[1:]) <= 16)
+    cfg.define_annotate('ann_reduce', [ib, kb, kh, kw], policy='try_unroll')
+
+    re_axes = cfg.define_reorder("reorder_0",
+                                 [n, co, oh, ow, vc, vh, vw, kh, kw, kb, ib, ci],
+                                 policy='interval_all', interval=(6, 11))
+    cfg.add_flop(2 * N * OH * OW * CO * CI * 8 * KH * KW)
```
Comment (eqy) [inline topi/python/topi/nn/bitserial_conv2d.py]: ""

File: topi/tests/python/test_topi_bitserial_conv2d_rasp.py [hunk 4]
```diff
@@ -45,6 +44,9 @@ def test_bitserial_conv2d():
     verify_bitserial_conv2d_nhwc(1, in_size, ic, oc, k, stride, pad, 1, 1, False)
     verify_bitserial_conv2d_nhwc(1, in_size, ic, oc, k, stride, pad, 2, 1, False)
 
+    verify_bitserial_conv2d_nhwc(1, in_size, ic, oc, k, stride, pad, 1, 1, True)
```
Comment (eqy) [inline topi/tests/python/test_topi_bitserial_conv2d_rasp.py]: Is it possible to add a correctness check here for random or precomputed inputs? It looks like `generate_quantized_np` is no longer used.

File: topi/python/topi/arm_cpu/bitserial_conv2d.py (hunk 1 above)
Comment (cowanmeg) [reply topi/python/topi/arm_cpu/bitserial_conv2d.py]: Sadly no :( it gives an error about multiple output in the compute.

File: topi/tests/python/test_topi_bitserial_conv2d_rasp.py (hunk 4 above)
Comment (cowanmeg) [reply topi/tests/python/test_topi_bitserial_conv2d_rasp.py]: Added, but it will only run on an arm device.

**Comment by tqchen** [comment]
@ajtulloch @vinx13 please also help review the PR if you have time

File: topi/python/topi/x86/bitserial_conv2d.py [hunk 5]
```diff
@@ -1,74 +1,33 @@
...
+def topi_bitserial_conv2d_nhwc(*args, **kwargs):
+    args = deserialize_args(args)
+    C = bitserial_conv2d_nhwc(*args, **kwargs)
+    s = generic.nn.schedule_bitserial_conv2d_nhwc([C])
+    data = args[0]
+    kernel = args[1]
+    return s, [data, kernel, C]
+
+@autotvm.task.register("topi_arm_cpu_bitserial_conv_nchw")
+@autotvm.task.register("topi_x86_bitserial_conv_nchw")
```
Comment (vinx13) [inline topi/python/topi/x86/bitserial_conv2d.py]: what about moving task registration to `topi_integration` and renaming to `topi_bitserial_conv_nchw`?

**Comment by tqchen** [comment]
Thanks, @vinx13 @cowanmeg @eqy, this is now merged

**Pull request merged**